services/w3c/w3c-validation.service.js

```diff
@@ -0,0 +1,269 @@
+'use strict'
+const url = require('url')
+const Joi = require('@hapi/joi')
+const { optionalUrl } = require('../validators')
+const { BaseJsonService } = require('..')
+
+const schema = Joi.object({
+  url: optionalUrl.required(),
+  messages: Joi.array()
+    .required()
+    .items(Joi.object()),
+}).required()
+
+const html5Expression =
+  '^HTML\\s?,\\s?SVG\\s?1\\.1\\s?,\\s?MathML\\s?3\\.0(\\s?,\\s?((ITS\\s?2\\.0)|(RDFa\\s?Lite\\s?1\\.1)))?$'
+
+const html4Expression =
+  '^HTML\\s?4\\.01\\s?(Strict|Transitional|Frameset)\\s?,\\s?URL\\s?\\/\\s?XHTML\\s?1\\.0\\s?(Strict|Transitional|Frameset)\\s?,\\s?URL$'
+const xhtmlExpression =
+  '^(XHTML\\s?,\\s?SVG\\s?1\\.1\\s?,\\s?MathML\\s?3\\.0(\\s?,\\s?RDFa\\s?Lite\\s?1\\.1)?)|(XHTML\\s?1\\.0\\s?Strict\\s?,\\s?URL\\s?,\\s?Ruby\\s?,\\s?SVG\\s?1\\.1\\s?,\\s?MathML\\s?3\\.0)$'
+const svgExpression =
+  '^SVG\\s?1\\.1\\s?,\\s?URL\\s?,\\s?XHTML\\s?,\\s?MathML\\s?3\\.0$'
+const presetRegex = new RegExp(
+  `(${html5Expression})|(${html4Expression})|(${xhtmlExpression})|(${svgExpression})`,
+  'i'
+)
+const parserRegex = new RegExp(`(^html$)|(^xml$)|(^xmldtd$)`, 'i')
+const queryParamDecode = regExpression => (value, helpers) => {
+  if (!value) {
+    return ''
+  }
+
+  if (regExpression && !decodeURI(value).match(regExpression)) {
+    return helpers.error('regex')
+  }
+
+  // Return the value unchanged
+  return value
+}
+
+const queryParamSchema = Joi.object({
+  doc: optionalUrl.required(),
+  preset: Joi.string()
+    .allow('')
+    .custom(queryParamDecode(presetRegex), 'preset validation'),
```

Comment:
Let's move `preset` to a route parameter, since only one option can be specified and there's only ~4 options we need to support. I've added an inline suggestion on the route definition on this note

Reply:
moved the parser (default,html,xml,xmldtd) instead of the preset